File: sdks/kotlin/src/main/kotlin/ly/neptune/astro/alias/AliasClient.kt
package ly.neptune.astro.alias

import ly.neptune.astro.HttpEngine
import ly.neptune.astro.models.*

class AliasClient internal constructor(private val engine: HttpEngine) {

    suspend fun getProfile(alias: String): AliasProfile =
        engine.get("/alias/$alias")

    suspend fun getAccounts(alias: String): AliasAccountsResponse =
        engine.get("/alias/$alias/accounts")

    suspend fun deactivate(alias: String): AliasProfile =
        engine.post("/alias/$alias/deactivate")

    /** Bank-server only. UNKNOWN is a registry transport state, not availability. */
    suspend fun availability(alias: String): AliasAvailability =
        engine.get("/alias/$alias/availability")

    /** Bank-server only. Identity permanently retires the previous NPT name. */
    suspend fun rename(request: RenameAliasRequest): RenameAliasResult =
        engine.patch("/alias/rename", request)

    suspend fun resolve(alias: String): ResolveResult =
        engine.get("/identity/resolve", mapOf("alias" to alias))
}

File: sdks/flutter/lib/src/alias/alias_client.dart
import '../astro_http.dart';
import '../models/models.dart';

class AliasClient {
  final AstroHttpClient _http;
  const AliasClient(this._http);

  Future<AliasProfile> getProfile(String alias) async {
    final json = await _http.get('/alias/$alias');
    return AliasProfile.fromJson(json as Map<String, dynamic>);
  }

  Future<List<LinkedAccount>> getAccounts(String alias) async {
    final json = await _http.get('/alias/$alias/accounts');
    final list = (json as Map<String, dynamic>)['accounts'] as List;
    return list.map((e) => LinkedAccount.fromJson(e as Map<String, dynamic>)).toList();
  }

  Future<AliasProfile> deactivate(String alias) async {
    final json = await _http.post('/alias/$alias/deactivate');
    return AliasProfile.fromJson(json as Map<String, dynamic>);
  }

  /// Bank-server only. UNKNOWN is a registry transport state, not availability.
  Future<AliasAvailability> availability(String alias) async {
    final encoded = Uri.encodeComponent(alias);
    final json = await _http.get('/alias/$encoded/availability');
    return AliasAvailability.fromJson(json as Map<String, dynamic>);
  }

  /// Bank-server only. Identity permanently retires the previous NPT name.
  Future<RenameAliasResult> rename(RenameAliasRequest request) async {
    final json = await _http.patch('/alias/rename', request.toJson());
    return RenameAliasResult.fromJson(json as Map<String, dynamic>);
  }

  Future<ResolveResult> resolve(String alias) async {
    final json = await _http.get('/identity/resolve', params: {'alias': alias});
    return ResolveResult.fromJson(json as Map<String, dynamic>);
  }
}
